File: code/py/src/modular_api/core/request_pipeline_audit.py
from __future__ import annotations

from dataclasses import dataclass

from starlette.types import Scope

REQUEST_PIPELINE_AUDIT_STATE_KEY = "modular_pipeline_audit"


@dataclass
class ShortCircuitAuditEntry:
    plugin_id: str
    middleware_id: str
    slot: str


@dataclass
class RequestPipelineAuditState:
    short_circuit: ShortCircuitAuditEntry | None = None


def ensure_request_pipeline_audit(scope: Scope) -> RequestPipelineAuditState:
    state = scope.setdefault("state", {})
    audit = state.get(REQUEST_PIPELINE_AUDIT_STATE_KEY)
    if audit is None:
        audit = RequestPipelineAuditState()
        state[REQUEST_PIPELINE_AUDIT_STATE_KEY] = audit
    return audit


def set_short_circuit_candidate(scope: Scope, candidate: ShortCircuitAuditEntry) -> None:
    ensure_request_pipeline_audit(scope).short_circuit = candidate


def clear_short_circuit_candidate(scope: Scope, middleware_id: str) -> None:
    audit = ensure_request_pipeline_audit(scope)
    if audit.short_circuit is not None and audit.short_circuit.middleware_id == middleware_id:
        audit.short_circuit = None

File: code/dart/modular_api/lib/src/core/request_pipeline_audit.dart
import 'package:shelf/shelf.dart';

const requestPipelineAuditContextKey = 'modular.requestPipelineAudit';

class ShortCircuitAuditEntry {
  final String pluginId;
  final String middlewareId;
  final String slot;

  const ShortCircuitAuditEntry({
    required this.pluginId,
    required this.middlewareId,
    required this.slot,
  });
}

class RequestPipelineAuditState {
  ShortCircuitAuditEntry? shortCircuit;
}

RequestPipelineAuditState? requestPipelineAuditFrom(Request request) {
  return request.context[requestPipelineAuditContextKey] as RequestPipelineAuditState?;
}

void setShortCircuitCandidate(Request request, ShortCircuitAuditEntry candidate) {
  requestPipelineAuditFrom(request)?.shortCircuit = candidate;
}

void clearShortCircuitCandidate(Request request, String middlewareId) {
  final audit = requestPipelineAuditFrom(request);
  if (audit?.shortCircuit?.middlewareId == middlewareId) {
    audit!.shortCircuit = null;
  }
}